static/app/views/dashboards/dashboardLegendUtils.tsx

```diff
@@ -0,0 +1,204 @@
+  router: InjectedRouter;
+
+  constructor(props: Props) {
+    this.dashboard = props.dashboard;
+    this.location = props.location;
+    this.organization = props.organization;
+    this.router = props.router;
+  }
+
+  // Updates legend param when a legend selection has been changed
+  updateLegendQueryParam(selected: Record<string, boolean>, widget: Widget) {
+    const [dashboard, location, router] = [this.dashboard, this.location, this.router];
+    const widgets = dashboard ? dashboard.widgets : [];
+    let newLegendQuery: string[];
+    if (!location.query.unselectedSeries && widgets) {
+      newLegendQuery = widgets
+        .filter(dashboardWidget => this.widgetRequiresLegendUnselection(dashboardWidget))
+        .map(dashboardWidget => {
+          return dashboardWidget.id === widget.id
+            ? this.encodeLegendQueryParam(widget, selected)
+            : this.formatLegendDefaultQuery(dashboardWidget.id);
+        });
+
+      const thisWidgetWithReleasesWasSelected =
+        Object.values(selected).filter(value => value === false).length !== 1 &&
+        Object.keys(selected).includes(`Releases:${widget.id}`);
+
+      const thisWidgetWithoutReleasesWasSelected =
+        !Object.keys(selected).includes(`Releases:${widget.id}`) &&
+        Object.values(selected).filter(value => value === false).length === 1;
+
+      if (thisWidgetWithReleasesWasSelected || thisWidgetWithoutReleasesWasSelected) {
+        router.replace({
+          query: {
+            ...location.query,
+            unselectedSeries: newLegendQuery,
+          },
+        });
+      }
+    } else if (Array.isArray(location.query.unselectedSeries)) {
+      let isInQuery = false;
+      newLegendQuery = location.query.unselectedSeries.map(widgetLegend => {
+        if (widgetLegend.includes(widget.id!)) {
+          isInQuery = true;
+          // names of legend options are stored as seriesName:widgetId and are stored in
+          // query param as widgetId-seriesName-seriesName2-...
+          return this.encodeLegendQueryParam(widget, selected);
+        }
+        return widgetLegend;
+      });
+
+      !isInQuery
+        ? router.replace({
+            query: {
+              ...location.query,
+              unselectedSeries: [
+                ...location.query.unselectedSeries,
+                this.encodeLegendQueryParam(widget, selected),
+              ],
+            },
+          })
+        : router.replace({
+            query: {
+              ...location.query,
+              unselectedSeries: newLegendQuery,
+            },
+          });
+    } else {
+      router.replace({
+        query: {
+          ...location.query,
+          unselectedSeries: [
+            location.query.unselectedSeries,
+            this.encodeLegendQueryParam(widget, selected),
+          ],
+        },
+      });
+    }
+  }
+
+  // sets unselected legend options by the legend query param
+  getLegendUnselected(widget: Widget) {
```

Comment:
To match the comment from above, a name like `getWidgetSelectionState` that returns `LegendSelection` is a nice tidy name that matches the `setWidgetSelectionState`